rbx_types/src/attributes/mod.rs

```diff
@@ -0,0 +1,141 @@
+    String = 0x02,
+    Bool = 0x03,
+    Float32 = 0x05,
+    Float64 = 0x06,
+    UDim = 0x09,
+    UDim2 = 0x0A,
+    BrickColor = 0x0E,
+    Color3 = 0x0F,
+    Vector2 = 0x10,
+    Vector3 = 0x11,
+    NumberSequence = 0x17,
+    ColorSequence = 0x19,
+    NumberRange = 0x1B,
+    Rect = 0x1C,
+});
+
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
+pub enum AttributeError {
+    #[error("invalid value type: {0}")]
+    InvalidValueType(u8),
+
+    #[error("invalid brick color: {0}")]
+    InvalidBrickColor(u32),
+
+    #[error("invalid entry key")]
+    InvalidEntryKey,
+
+    #[error("invalid name")]
+    InvalidName,
+
+    #[error("invalid size")]
+    InvalidSize,
+
+    #[error("invalid variant type passed")]
+    InvalidVariantType,
+
+    #[error("no value type was found")]
+    NoValueType,
+
+    #[error("couldn't read bytes to deserialize {0}")]
+    Other(&'static str),
+}
+
+#[cfg(all(feature = "serde", test))]
+mod tests {
+    use super::*;
+
+    // This is taken from rbx-test-files/models/attributes/xml.rbxmx.
+    // This is pasted raw as to not create a circular dependency in test (rbx_types -> rbx_xml/rbx_binary -> rbx_types)
+    const ATTRIBUTES_BASE64: &str = r"
+    DwAAAAMAAABOYU4GAAAAAAAA+P8IAAAASW5maW5pdHkGAAAAAAAA8H8NAAAAQ29sb3JTZXF1
+    ZW5jZRkDAAAAAAAAAAAAAAAAAIA/AAAAAAAAAAAAAAAAAAAAPwAAAAAAAIA/AAAAAAAAAAAA
+    AIA/AAAAAAAAAAAAAIA/BwAAAFZlY3RvcjMRAACAPwAAAEAAAEBABwAAAFZlY3RvcjIQAAAg
+    QQAASEIOAAAATnVtYmVyU2VxdWVuY2UXAwAAAAAAAAAAAAAAAACAPwAAAAAAAAA/AAAAAAAA
+    AAAAAIA/AACAPwYAAABDb2xvcjMPo6IiPwAAAAAAAIA/CgAAAEJyaWNrQ29sb3IO7AMAAAQA
+    AABSZWN0HAAAgD8AAABAAABAQAAAgEAFAAAAVURpbTIKAAAAPwoAAAAzMzM/HgAAAAQAAABV
+    RGltCQAAAD9kAAAACwAAAE51bWJlclJhbmdlGwAAoEAAACBBBgAAAE51bWJlcgYAAAAAgBzI
+    QAcAAABCb29sZWFuAwEGAAAAU3RyaW5nAg0AAABIZWxsbywgd29ybGQh
+    ";
+
+    #[test]
+    fn test_attributes() {
+        let attributes_value =
+            base64::decode(&ATTRIBUTES_BASE64.split_whitespace().collect::<String>())
+                .expect("bad base64 for attributes");
+
+        let attributes =
+            get_attributes(&attributes_value[..]).expect("couldn't deserialize attributes");
+
+        let mut attributes_stable_order = attributes.into_iter().collect::<Vec<_>>();
+        attributes_stable_order.sort_by_cached_key(|(key, _)| key.to_owned());
+        insta::assert_yaml_snapshot!(attributes_stable_order);
+
+        let new_attribute_bytes = attributes_from_map(attributes_stable_order.to_owned())
+            .expect("couldn't get attributes from map");
+        let new_attributes = get_attributes(&new_attribute_bytes[..])
+            .expect("couldn't deserialize crate produced binary");
+        let mut new_attributes_stable_order = new_attributes.into_iter().collect::<Vec<_>>();
+        new_attributes_stable_order.sort_by_cached_key(|(key, _)| key.to_owned());
+
+        // They are not checked directly against each other because the data contains NaN.
+        assert_eq!(
```

Comment:
We should probably be using insta snapshots here instead of just manually asserting that these are equal.

Reply:
This would be an identical snapshot with `attributes_stable_order` in a working case, which is already in its own snapshot. The `assert_eq!` is there because that way, it's impossible to flub it and have two separate snaps.

Reply:
Seems reasonable.